tensorflow/contrib/android/java/org/tensorflow/contrib/android/TensorFlowInferenceInterface.java

```diff
@@ -46,9 +53,99 @@
  * example usage.
  */
 public class TensorFlowInferenceInterface {
-  private static final String TAG = "TensorFlowInferenceInterface";
+  private static final String TAG = "TensorFlowInferenceInt";
   private static final String ASSET_FILE_PREFIX = "file:///android_asset/";
 
+  /**
+   * Copies raw resource to a cache file.
+   *
+   * @return File reference to cache file.
+   */
+  private File createCacheFile(Context context, int resourceId, String filename)
+      throws IOException {
+    File cacheFile = new File(context.getCacheDir(), filename);
+
+    if (cacheFile.createNewFile() == false) {
```

Comment:
If the file is already present, you could check that size and timestamp of the file matches the one from assets, and if so just avoid copying altogether.

Reply:
Remember I copy from res/raw not assets. It is faster that way but still suffers from the fact that aapt compresses it.  To get its original size for comparison I have to copy it again which defeats the purpose of this check.  I delete the file at the end.  It is just for safety's sake that the file check is there in case somehow another file got written with the same name.

Reply:
What is the advantage of res/raw over assets/?

It's possible to explicitly keep aapt from compressing extensions, e.g. with [noCompress](https://google.github.io/android-gradle-dsl/current/com.android.build.gradle.internal.dsl.AaptOptions.html#com.android.build.gradle.internal.dsl.AaptOptions:noCompress) in Gradle. I think passing the --uncompressedExtensions flag to bazel might do the same thing.

A properly trained model will probably be very high entropy anyway, so typically you won't see a lot of compression savings. And in fact I just checked the demo -- found that none of the .pb files in it are compressed at all, simply stored. In that case, can we just mmap it from assets/ directly without the copy?

Reply:
res/raw is indexed in the R file and claims on SO to be faster.  I did not benchmark it to prove it.
Of course I could not get a map with compression so the assets would *have* to be uncompressed, which might be hard to enforce.
Right now I am trying to test if a filedescriptor can be used on uncompressed asset to make a FileInputStream--the only way I have not yet tried without copy.

Reply:
res/raw is indexed in the R file and claims on SO to be faster. It does come back faster than the assets.
Of course I could not get a memory map with compression so the assets would have to be uncompressed for a direct map, which unfortunately, creates an OutOfMemoryError, when trying to open from a FileInputStream. The (not retrained) inception file is 50MB which is half your allowance on Android, for instance. It makes little sense to try to do it that way for this reason. People want to use it in apps, not just stand alone and will run out of memory fast.
I will push the file again with the other corrections as per above

Reply:
My major reservation is that this commit only seems to speed up loading from res. It might be nice to support this functionality, but up until this point we've suggested assets/ as the place to place pb files. So it will be a bit more difficult for users to switch to the new method, and possibly confuse things unless well documented. How much faster is it actually to copy from res? Also, why does it give you an OOM error when mapping from the cache dir, but not from assets/?

It should be possible to mmap files directly from assets with the nocompressextensions directive in bazel. In which case the copy is not even necessary. Can you tell if this is the case?